### PR TITLE
Use FI_EFA_FORK_SAFE environment variable to enable fork safety support

### PR DESCRIPTION
**Pull request by nzmsv**
This makes use of compatibility checks in Libfabric and only enables
the feature when it is needed.  When running with a newer kernel with
copy-on-fork support and recent rdma-core, additional fork safety
support is not required and so the variable is silently ignored.

Using FI_EFA_FORK_SAFE instead of FI_EFA_FORK_SAFE does the right
thing on both new and legacy systems.

Signed-off-by: Greg Inozemtsev <ginozemt@amazon.com>

*Issue #, if available:*
NCCLOFI-273

*Description of changes:*
Version of #82 for `master`.  Tested on P4dn with all providers.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nzmsv** [comment]
This was tested with Libfabric before and after the addition of the `FI_EFA_FORK_SAFE` variable, just like #82.

File: src/nccl_ofi_net.c [hunk 1]
```diff
@@ -992,9 +992,17 @@ static ncclResult_t ofi_init(ncclDebugLogger_t logFunction)
 	NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using " PACKAGE_STRING);
 
 	/*
-	 * RDMAV_FORK_SAFE environment variable makes the rdma-core
-	 * library fork-safe. This significantly increases cost of memory
-	 * registration when huge pages are enabled.
+	 * FI_EFA_FORK_SAFE environment variable tells Libfabric to enable
+	 * fork-safe support in legacy versions of the rdma-core library.
+	 * This significantly increases cost of memory registration. Libfabric
```
Comment (rashikakheria) [inline src/nccl_ofi_net.c:997]: How does this increase memory registration cost when huge pages get disabled inside libfabric?

File: src/nccl_ofi_net.c [hunk 2]
```diff
@@ -1003,14 +1011,19 @@ static ncclResult_t ofi_init(ncclDebugLogger_t logFunction)
...
 	 * To prevent NCCL based applications from getting aborted when using
-	 * fork(), plugin explicitly enables RDMAV_FORK_SAFE environment
-	 * variable.
+	 * fork(), the plugin explicitly enables FI_EFA_FORK_SAFE environment
+	 * variable, even in legacy environments where the overhead is high.
 	 */
-	if (!getenv("RDMAV_FORK_SAFE")) {
-		NCCL_OFI_INFO(NCCL_INIT, "Setting RDMAV_FORK_SAFE environment variable to 1.");
-		rc = setenv("RDMAV_FORK_SAFE", "1", 1);
+	const uint32_t libversion = fi_version();
```
Comment (rashikakheria) [inline src/nccl_ofi_net.c:1017]: Does it make sense to have this as a global variable for any future libfabric compatibility changes?

File: src/nccl_ofi_net.c [hunk 3]
```diff
@@ -1003,14 +1011,19 @@ static ncclResult_t ofi_init(ncclDebugLogger_t logFunction)
...
+	const char * fork_safe_var_name =
+		(FI_MAJOR(libversion) > 1 || (FI_MAJOR(libversion) == 1 && FI_MINOR(libversion) >= 13))
+		? "FI_EFA_FORK_SAFE"
+		: "RDMAV_FORK_SAFE";
+	if (!getenv(fork_safe_var_name)) {
```
Comment (rashikakheria) [inline src/nccl_ofi_net.c:1022]: You might want to handle env variable set to 0 as you handled in similar patches to libfabric and rdma-core.

**Review by rashikakheria — CHANGES_REQUESTED** [review]

File: src/nccl_ofi_net.c (hunk 3 above)
Comment (nzmsv) [reply src/nccl_ofi_net.c:1022]: Same thinking as my comment on #82: if the user explicitly wants to disable this, should we let them?

File: src/nccl_ofi_net.c (hunk 1 above)
Comment (nzmsv) [reply src/nccl_ofi_net.c:997]: This was just copied from the earlier version of the comment.  I removed this sentence.

**Review by rashikakheria — APPROVED** [review]

**Review by rashikakheria — APPROVED** [review]

File: RELEASENOTES.md [hunk 4]
```diff
@@ -7,6 +7,29 @@
...
+maintaining backward compatibility with older NCCL versions (up to [NCCL v2.4.8](https://github.com/NVIDIA/nccl/releases/tag/v2.4.8-1)).
+It was tested with Libfabric versions up to [Libfabric v1.14.0](https://github.com/ofiwg/libfabric/releases/tag/v1.14.0).
+
+New Features:
+* Make use of FI_EFA_FORK_SAFE environment variable to allow Libfabric to detect when `MADV_DONTFORK`
+  is not needed (#82).  This feature requires Libfabric v1.13.0 or higher.  When used with an older version
+  of Libfabric, the plugin will continue to set the RDMAV_FORK_SAFE environment variable.
+
+Bug Fixes:
+* Ensure that the buffer used for flush is page aligned and allocated with `mmap` instead of `malloc`.
```
Comment (rashikakheria) [inline RELEASENOTES.md:23]: There are other bug fixes that have been merged to branches. Please include that in release notes.

**Review by rashikakheria — CHANGES_REQUESTED** [review]

**Comment by nzmsv** [comment]
I decided to remove release notes from this change and make a separate PR.

**Review by rashikakheria — APPROVED** [review]

**Pull request merged**